### PR TITLE
feat: add typehint on adapter

### PR DESCRIPTION
**Pull request by KevinArtus**

### CODE REVIEW

File: src/Gaufrette/Adapter/AwsS3.php [hunk 1]
```diff
@@ -182,18 +173,29 @@ public function size($key)
         }
     }
 
+    public function mimeType(string $key): string
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/AwsS3.php:176]:
```suggestion
    public function mimeType(string $key): string|bool
```

File: src/Gaufrette/Adapter/AwsS3.php [hunk 2]
```diff
@@ -307,22 +309,13 @@ protected function computePath($key)
...
         return ltrim(substr($path, strlen($this->options['directory'])), '/');
     }
 
-    /**
-     * @param string $content
-     *
-     * @return string
-     */
-    private function guessContentType($content)
+    private function guessContentType(mixed $content): false|string
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/AwsS3.php:318]: not available on 8.0.
```suggestion
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 3]
```diff
@@ -177,10 +142,8 @@ public function read($key)
 
     /**
      * {@inheritdoc}
-     * @throws \RuntimeException
-     * @throws \InvalidArgumentException
      */
-    public function write($key, $content)
+    public function write(string $key, mixed $content): int|bool
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:146]: same, why not `string $content` ?

File: src/Gaufrette/Adapter/Flysystem.php [hunk 4]
```diff
@@ -6,24 +6,17 @@
...
+    private AdapterInterface $adapter;
+    private Config $config;
 
     /**
-     * @param AdapterInterface  $adapter
-     * @param \League\Flysystem\Config|array|null $config
+     * @param Config|array|null $config
      */
-    public function __construct(AdapterInterface $adapter, $config = null)
+    public function __construct(AdapterInterface $adapter, Config $config = null)
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/Flysystem.php:19]:
```suggestion
    public function __construct(AdapterInterface $adapter, ?Config $config = null)
```

Comment (Zusoy) [inline src/Gaufrette/Adapter/Flysystem.php:19]:
```suggestion
    public function __construct(AdapterInterface $adapter, ?Config $config = null)
```

File: src/Gaufrette/Adapter/PhpseclibSftp.php [hunk 5]
```diff
@@ -56,7 +59,7 @@ public function rename($sourceKey, $targetKey)
     /**
      * {@inheritdoc}
      */
-    public function write($key, $content)
+    public function write(string $key, mixed $content): int|bool
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/PhpseclibSftp.php:62]: Same, wy not `string $content` ?

File: src/Gaufrette/Adapter/PhpseclibSftp.php [hunk 6]
```diff
@@ -184,7 +188,7 @@ protected function initialize()
         $this->initialized = true;
     }
 
-    protected function ensureDirectoryExists($directory, $create)
+    protected function ensureDirectoryExists(string $directory, bool $create)
```
Comment (Zusoy) [inline src/Gaufrette/Adapter/PhpseclibSftp.php:191]:
```suggestion
    protected function ensureDirectoryExists(string $directory, bool $create): void
```

File: src/Gaufrette/Adapter.php [hunk 7]
```diff
@@ -13,72 +13,48 @@ interface Adapter
...
     /**
      * Writes the given content into the file.
      *
-     * @param string $key
-     * @param string $content
-     *
      * @return int|bool The number of bytes that were written into the file
      */
-    public function write($key, $content);
+    public function write(string $key, mixed $content): int|bool;
```
[the remark this answers is not in the document]
Comment (KevinArtus) [reply src/Gaufrette/Adapter.php:25]: This class is an interface implemented by lot of classes.
Those class implements write which will call methods where content is used as a resource

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 8]
```diff
@@ -49,14 +43,12 @@ public function __construct(SimpleS3Client $service, $bucket, array $options = [
...
     }
 
     /**
      * {@inheritdoc}
      */
-    public function setMetadata($key, $content)
+    public function setMetadata(string $key, array $content): void
     {
         // BC with AmazonS3 adapter
         if (isset($content['contentType'])) {
```
[the remark this answers is not in the document]
Comment (KevinArtus) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:54]: Because isset will also test if the data is not null. Not `array_key_exists`

File: src/Gaufrette/Adapter/AwsS3.php [hunk 9]
```diff
@@ -47,14 +40,12 @@ public function __construct(S3Client $service, $bucket, array $options = [], $de
...
     }
 
     /**
      * {@inheritdoc}
      */
-    public function setMetadata($key, $metadata)
+    public function setMetadata(string $key, array $content): void
     {
         // BC with AmazonS3 adapter
         if (isset($metadata['contentType'])) {
```
[the remark this answers is not in the document]
Comment (KevinArtus) [reply src/Gaufrette/Adapter/AwsS3.php:51]: same

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 10]
```diff
@@ -10,37 +10,24 @@
...
-     * @param bool     $detectContentType
-     */
-    public function __construct(SimpleS3Client $service, $bucket, array $options = [], $detectContentType = false)
-    {
+    protected array $options;
+    protected bool $bucketExists;
+    protected array $metadata = [];
+
+    public function __construct(
+        private readonly SimpleS3Client $service,
```
Comment (muchafm) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:23]: @KevinArtus we said that we will add the property promotion et readonly once the typehint will be merge right ?

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 11]
```diff
@@ -314,22 +305,13 @@ protected function computePath($key)
...
         return ltrim(substr($path, strlen($this->options['directory'])), '/');
     }
 
-    /**
-     * @param string|resource $content
-     *
-     * @return string
-     */
-    private function guessContentType($content)
+    private function guessContentType(mixed $content): false|string
```
[the remark this answers is not in the document]
Comment (muchafm) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:307]:
```suggestion
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Adapter/AwsS3.php [hunk 12]
```diff
@@ -9,36 +9,25 @@
...
-    public function __construct(S3Client $service, $bucket, array $options = [], $detectContentType = false)
-    {
+    protected S3Client $service;
+    protected array $options;
+    protected bool $bucketExists;
+    protected array $metadata = [];
+
+    public function __construct(
+        S3Client $service,
+        private readonly string $bucket,
```
Comment (muchafm) [inline src/Gaufrette/Adapter/AwsS3.php:23]: same here

**Review by muchafm — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 13]
```diff
@@ -10,37 +10,31 @@
...
-    public function __construct(SimpleS3Client $service, $bucket, array $options = [], $detectContentType = false)
-    {
+    protected SimpleS3Client $service;
+    protected string $bucket;
+    protected array $options;
+    protected bool $detectContentType;
+    protected bool $bucketExists;
+    protected array $metadata = [];
+
+    public function __construct(
```
[the remark this answers is not in the document]
Comment (PedroTroller) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:26]: Not the goal of these PR, will be done in an other.

File: src/Gaufrette/Adapter/AsyncAwsS3.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (PedroTroller) [reply src/Gaufrette/Adapter/AsyncAwsS3.php:57]: And... it's not the goal of there PR :kissing:

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (KevinArtus) [reply src/Gaufrette/Adapter.php:25]: For example, the method `write` on `Gaufrette\Adapter\AzureBlobStorage` calls the foo method `guessContentType`.

This needs a resource to execute its code

File: src/Gaufrette/Adapter/Ftp.php [hunk 14]
```diff
@@ -582,14 +551,17 @@ private function connect()
...
+    public function close(): void
     {
         if ($this->isConnected()) {
             ftp_close($this->connection);
         }
     }
 
-    private function isLinuxListing($info)
+    /**
+     * @param array<int, string>|false $info
```
Comment (muchafm) [inline src/Gaufrette/Adapter/Ftp.php:562]:
```suggestion
     * @param array<int, string>|bool $info
```

File: src/Gaufrette/StreamWrapper.php [hunk 15]
```diff
@@ -58,149 +56,129 @@ public static function register($scheme = 'gaufrette')
...
+     * @param string $scheme Protocol scheme
      */
-    protected static function streamWrapperRegister($scheme, $className)
+    protected static function streamWrapperRegister(string $scheme, string $className): bool
     {
         return stream_wrapper_register($scheme, $className);
     }
 
-    public function stream_open($path, $mode)
+    /**
```
Comment (muchafm) [inline src/Gaufrette/StreamWrapper.php:84]: maybe add a bit of explanation for the number 9 ?

**Review by muchafm — CHANGES_REQUESTED** [review]

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 16]
```diff
@@ -129,12 +90,9 @@ public function createContainer($containerName, CreateContainerOptions $options
     /**
      * Deletes a container.
      *
-     * @param string                 $containerName
-     * @param BlobServiceOptions $options
-     *
      * @throws \RuntimeException if cannot delete the container
      */
-    public function deleteContainer($containerName, BlobServiceOptions $options = null)
+    public function deleteContainer(string $containerName, BlobServiceOptions $options = null)
```
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:95]:
```suggestion
    public function deleteContainer(string $containerName, BlobServiceOptions $options = null): void
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 17]
```diff
@@ -97,12 +61,9 @@ public function setCreateContainerOptions(CreateContainerOptions $options)
     /**
      * Creates a new container.
      *
-     * @param string                                                     $containerName
-     * @param \MicrosoftAzure\Storage\Blob\Models\CreateContainerOptions $options
-     *
      * @throws \RuntimeException if cannot create the container
      */
-    public function createContainer($containerName, CreateContainerOptions $options = null)
+    public function createContainer(string $containerName, CreateContainerOptions $options = null)
```
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:66]:
```suggestion
    public function createContainer(string $containerName, CreateContainerOptions $options = null): void
```

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 18]
```diff
@@ -28,67 +29,30 @@ class AzureBlobStorage implements Adapter, MetadataSupporter, SizeCalculator, Ch
...
-    public function getCreateContainerOptions()
+    public function getCreateContainerOptions(): CreateContainerOptions
     {
         return $this->createContainerOptions;
     }
 
-    /**
-     * @param CreateContainerOptions $options
-     */
     public function setCreateContainerOptions(CreateContainerOptions $options)
```
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AzureBlobStorage.php:56]:
```suggestion
    public function setCreateContainerOptions(CreateContainerOptions $options): void
```

File: src/Gaufrette/Adapter/Ftp.php [hunk 19]
```diff
@@ -475,20 +448,16 @@ private function parseRawlist(array $rawlist)
 
     /**
      * Computes the path for the given key.
-     *
-     * @param string $key
      */
-    private function computePath($key)
+    private function computePath(string $key)
```
Comment (KevinArtus) [inline src/Gaufrette/Adapter/Ftp.php:452]:
```suggestion
    private function computePath(string $key): string
```

File: src/Gaufrette/Adapter/Ftp.php (hunk 14 above)
Comment (PedroTroller) [reply src/Gaufrette/Adapter/Ftp.php:562]: Event with false... this method is completly fuck-up !

**Review by PedroTroller — CHANGES_REQUESTED** [review]

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 20]
```diff
@@ -314,22 +308,13 @@ protected function computePath($key)
...
         return ltrim(substr($path, strlen($this->options['directory'])), '/');
     }
 
-    /**
-     * @param string|resource $content
-     *
-     * @return string
-     */
-    private function guessContentType($content)
+    private function guessContentType(mixed $content): bool|string
```
Comment (PedroTroller) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:317]:
```suggestion   
   /**
     * @param string|resource $content
     */
    private function guessContentType(mixed $content): bool|string
```

File: src/Gaufrette/Stream/InMemoryBuffer.php [hunk 21]
```diff
@@ -148,15 +145,15 @@ public function flush()
...
     {
         return $this->position >= $this->numBytes;
     }
 
     /**
-     * {@inheritdoc}
+     * @return array<string, mixed>|false
      */
-    public function stat()
+    public function stat(): array|false
```
Comment (KevinArtus) [inline src/Gaufrette/Stream/InMemoryBuffer.php:156]:
```suggestion
    public function stat(): array|bool
```

File: src/Gaufrette/Stream/Local.php [hunk 22]
```diff
@@ -53,10 +50,7 @@ public function open(StreamMode $mode)
         return true;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function read($count)
+    public function read(int $count): string|false
```
Comment (KevinArtus) [inline src/Gaufrette/Stream/Local.php:53]:
```suggestion
    public function read(int $count): string|bool
```

File: src/Gaufrette/Stream/Local.php [hunk 23]
```diff
@@ -153,9 +127,9 @@ public function eof()
     }
 
     /**
-     * {@inheritdoc}
+     * @return array<string, mixed>|false
      */
-    public function stat()
+    public function stat(): array|false
```
Comment (KevinArtus) [inline src/Gaufrette/Stream/Local.php:132]:
```suggestion
    public function stat(): array|bool
```

File: src/Gaufrette/Adapter/AsyncAwsS3.php [hunk 24]
```diff
@@ -10,37 +10,32 @@
...
-     * @param bool     $detectContentType
-     */
-    public function __construct(SimpleS3Client $service, $bucket, array $options = [], $detectContentType = false)
-    {
+    protected SimpleS3Client $service;
+    protected string $bucket;
+    protected array $options;
+    protected bool $bucketExists;
+    protected array $metadata = [];
+
```
Comment (KevinArtus) [inline src/Gaufrette/Adapter/AsyncAwsS3.php:23]:
```suggestion
```

**Review by muchafm — APPROVED** [review]
